doc/developer-guide/io-framework.md

```diff
@@ -0,0 +1,294 @@
+
+The changes required to use this framework can be significant given that
+it's based on a callback architecture while current implementation is
+basically sequential. For this reason it will be very useful that the
+framework can fully replace the current code to avoid maintaining two
+very different implementations, even if the legacy implementation is used.
+
+For example, in the current implementation there are two supported ways to
+do I/O:
+
+- **Synchronous I/O** _(legacy implementation)_
+  This is the most simple approach. Every time a I/O operation is done,
+  it's executed in the foreground, blocking the executing thread until
+  the operation finishes.
+
+- **io_uring I/O** ^[1]
+  This is a new and powerful kernel API that provides asynchronous I/O
+  execution with little overhead. This approach is superior because it
+  doesn't block the executing thread, allowing more work to be done while
+  the I/O operation is being processed in the background.
+
+- **threaded I/O**
+  This mode is not yet implemented, but it should be a replacement for the
+  legacy mode but with the advantages of the thread pool of the io_uring
+  approach. This means that all I/O will be done in background without
+  blocking the worker threads.
+
+_io_uring_ is only present on latest linux kernels and it's dynamically
+detected and used if available. Otherwise it silently fails back to the
+synchronous implementation in a transparent way for the rest of the code
+that uses this framework (once implemented, it will use the threaded I/O
+instead of the legacy mode when io_uring is not supported).
+
+The implementation is done with the io_uring API in mind. This means that
+io_uring fits very well in the I/O framework API, and the other modes
+are adjusted to follow the same semantics.
+
+## How to use it
+
+In this section a general overview of the operation will be provided,
+focused on the io_uring-based implementation. For differences when
+io_uring is not present, check section [Fallback mode](#fallback-mode)
+
+### Initialization
+
+The framework is initialized using `gf_io_run()`.
+
+```c
+typedef int32_t (*gf_io_async_t)(gf_io_op_t *op);
+
+typedef struct {
+    gf_io_async_t setup;
+    gf_io_async_t cleanup;
+} gf_io_handlers_t;
+
+int32_t gf_io_run(gf_io_handlers_t handlers, void *data);
+```
+
+The handlers structure contains two async functions, one that is called
+just after having initialized the I/O infrastructure, and another one that
+is called after stopping everything else. The 'data' argument is an extra
+argument that will be passed to each function.
+
+The returned value can be a negative error code if there has been any
+problem while initializing the system, or 0 if everything worked fine. In
+this case, the function only returns when the program is terminating.
+
+### Termination
+
+When it's determined that the process must be terminated, a call to
+`gf_io_shutdown()` must be done.
```

Comment:
The full implementation is not still there because it will require changes in other places that are not done right now. For this particular case, io_uring engine should work on it's own. However it currently depends on the events framework, which relies on `gf_event_dispatch()` as the core function. With this dependency it's hard to properly implement shutdown procedure.

After this patch, other components will need to be migrated incrementally to use this I/O framework. Once the basic parts are migrated, this will be implemented.

Anyway, I'll need to update the documentation to add more details and explain all these things, so that the documentation also progresses incrementally.